File: 2026-04/2026-04-19/unique_stair_climber.py
def get_unique_climbs(steps):
    if steps <= 2:
        return steps

    ways_to_climb_up_to_two_steps_before = 1
    ways_to_climb_up_to_one_step_before = 2
    current_count = 0

    for i in range(3, steps + 1):
        current_count = ways_to_climb_up_to_two_steps_before + ways_to_climb_up_to_one_step_before

        ways_to_climb_up_to_two_steps_before = ways_to_climb_up_to_one_step_before
        ways_to_climb_up_to_one_step_before = current_count

    return current_count

# --- TEST SUITE ---

tests_text = r'''
1. get_unique_climbs(4) should return 5.
2. get_unique_climbs(5) should return 8.
3. get_unique_climbs(10) should return 89.
4. get_unique_climbs(18) should return 4181.
5. get_unique_climbs(29) should return 832040.
6. get_unique_climbs(50) should return 20365011074.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 2026-04/2026-04-19/uniqueStairClimber.js
function getUniqueClimbs(steps) {

/* Backtracking solution that works for small number of steps (e.g. 4, 5, 10, 18) but takes long and fills the stack for larger numbers
  let count = 0;

  const backtrack = stepNumber => {
    if (stepNumber >= steps) {
      if (stepNumber === steps) {
        ++count;
      }
      
      return;
    }

    for (const stepsAtATimeChoice of [1, 2]) {
      stepNumber += stepsAtATimeChoice;
      backtrack(stepNumber);
      stepNumber -= stepsAtATimeChoice;
    }
  }

  backtrack(0);

  return count;
*/

// Solution that keeps a running tally using only the last two steps counts
  if (steps <= 2) {
    return steps;
  }

  let waysToClimbUpToTwoStepsBefore = 1;
  let waysToClimbUpToOneStepBefore = 2;
  let currentCount = 0;

  for (let i = 3; i <= steps; ++i) {
    currentCount = waysToClimbUpToTwoStepsBefore + waysToClimbUpToOneStepBefore;

    waysToClimbUpToTwoStepsBefore = waysToClimbUpToOneStepBefore;
    waysToClimbUpToOneStepBefore = currentCount;
  }

  return currentCount;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. getUniqueClimbs(4) should return 5.
2. getUniqueClimbs(5) should return 8.
3. getUniqueClimbs(10) should return 89.
4. getUniqueClimbs(18) should return 4181.
5. getUniqueClimbs(29) should return 832040.
6. getUniqueClimbs(50) should return 20365011074.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length) return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i])) {
          return false;
        }
      } else {
        return false;
      }
    } else if (a[i] !== b[i]) {
      return false;
    }
  }

  return true;
}

runTests(testData);